src/bio/bio_device.c

```diff
@@ -688,115 +812,71 @@ bio_set_led_state(struct bio_xs_context *xs_ctxt, uuid_t dev_uuid,
-			bio_dev->bb_name);
-		return -DER_INVAL;
-	}
-
-
-	if (spdk_pci_addr_parse(&pci_addr, b_info.bdi_traddr)) {
-		D_ERROR("Unable to parse PCI address: %s\n", b_info.bdi_traddr);
-		D_GOTO(free_traddr, rc = -DER_INVAL);
-	}
-
-	for (pci_device = spdk_pci_get_first_device(); pci_device != NULL;
-	     pci_device = spdk_pci_get_next_device(pci_device)) {
-		if (spdk_pci_addr_compare(&pci_addr, &pci_device->addr) == 0) {
-			found = true;
-			break;
-		}
+	if (opts.led_state == SPDK_VMD_LED_STATE_UNKNOWN) {
+		D_ERROR("LED state is not valid or supported\n");
+		return -DER_OP_UNEXP_RESULT;
 	}
 
-	if (found) {
-		if (strcmp(spdk_pci_device_get_type(pci_device), "vmd") != 0) {
-			D_DEBUG(DB_MGMT, "%s is not a VMD device\n",
-				b_info.bdi_traddr);
-			D_GOTO(free_traddr, rc = -DER_NOSYS);
-		}
-	} else {
-		D_ERROR("Unable to set led state, VMD device not found\n");
-		D_GOTO(free_traddr, rc = -DER_INVAL);
+	rc = fill_in_traddr(&b_info, bio_dev->bb_name);
+	if (rc != 0) {
+		D_ERROR("Unable to get traddr for device %s\n", bio_dev->bb_name);
+		return -DER_OP_FAILED;
 	}
 
-	/* First check the current state of the VMD LED */
-	rc = spdk_vmd_get_led_state(pci_device, &current_led_state);
-	if (rc) {
-		D_ERROR("Failed to get the VMD LED state\n");
+	rc = spdk_pci_addr_parse(&opts.pci_addr, b_info.bdi_traddr);
+	if (rc != 0) {
+		D_ERROR("Unable to parse PCI address for device %s (%s)\n", b_info.bdi_traddr,
+			spdk_strerror(-rc));
 		D_GOTO(free_traddr, rc = -DER_INVAL);
 	}
 
-	/* If the current state of a device is FAULTY we do not want to reset */
-	if ((current_led_state == SPDK_VMD_LED_STATE_FAULT) && reset)
-		D_GOTO(state_set, rc);
-
-	if (!reset)
-		D_DEBUG(DB_MGMT, "Setting VMD device:%s LED state to %s(%d)\n",
-			b_info.bdi_traddr, led_state, new_led_state);
-	else
-		D_DEBUG(DB_MGMT, "Resetting VMD device:%s LED state to %d\n",
-			b_info.bdi_traddr, bio_dev->bb_led_state);
-
-	/* Save the current state in bio_bdev, will be restored by init xs */
-	if (!reset)
-		bio_dev->bb_led_state = current_led_state;
-
-	if (current_led_state == new_led_state)
-		D_GOTO(state_set, rc);
+	spdk_pci_for_each_device(&opts, led_device_action);
 
-	/* Set the LED to the new state */
-	rc = spdk_vmd_set_led_state(pci_device, new_led_state);
-	if (rc) {
-		D_ERROR("Failed to set LED state to %s\n", led_state);
-		D_GOTO(free_traddr, rc = -DER_INVAL);
+	if (opts.status != 0) {
+		D_ERROR("Setting LED state failed\n");
+		D_GOTO(free_traddr, rc = opts.status);
 	}
-
-	rc = spdk_vmd_get_led_state(pci_device, &current_led_state);
-	if (rc) {
-		D_ERROR("Failed to get the VMD LED state\n");
-	} else {
-		/* Verify the correct state is set */
-		if (current_led_state != new_led_state)
-			D_ERROR("LED of device:%s is in an unexpected state:"
-				"%d\n", b_info.bdi_traddr, current_led_state);
+	if (!opts.all_devices && !opts.finished) {
+		D_ERROR("Device with traddr %s could not be found\n", b_info.bdi_traddr);
+		D_GOTO(free_traddr, rc = opts.status);
```

Comment:
If opts.status is != 0 then the code will not reach this point, so this code-path is logging an error then returning 0.  Perhaps -DER_INVAL would be better here?
```suggestion
		D_GOTO(free_traddr, rc = -DER_INVAL);
```

Reply:
done